src/coreclr/jit/llvm.cpp

```diff
@@ -1652,33 +1708,43 @@ Value* Llvm::getLocalVarAddress(GenTreeLclVar* lclVar) {
     return _builder.CreateGEP(_function->getArg(0), _builder.getInt32(varOffset), "lclVar");
 }
 
+bool Llvm::isLlvmFrameLocal(LclVarDsc* varDsc)
+{
+    assert(canStoreLocalOnLlvmStack(varDsc) && (_compiler->fgSsaPassesCompleted >= 1));
+    return !varDsc->lvInSsa;
+}
+
 void Llvm::storeLocalVar(GenTreeLclVar* lclVar)
 {
-    if (lclVar->gtFlags & GTF_VAR_DEF)
+    Value* localValue = getGenTreeValue(lclVar->gtGetOp1());
```

Comment:
Done.

Reply:
Ah, this is hitting an assert.  I'll need to see what that is about.

Reply:
I think it needs to handle `TYP_STRUCT`s.

Reply:
Maybe structs are also a problem, but this one is a phi bool:
```
------------ BB08 [099..09E) -> BB09 (always), preds={BB06,BB07} succs={BB09}
     (  0,  0) [000366] ------------       t366 =    PHI_ARG   bool   V00 loc0         u:8
     (  0,  0) [000365] ------------       t365 =    PHI_ARG   bool   V00 loc0         u:6
                                                  /--*  t366   bool
                                                  +--*  t365   bool
               [000357] ------------       t357 = *  PHI       bool
                                                  /--*  t357   bool
               [000358] DA----------              *  STORE_LCL_VAR bool   V00 loc0         d:7
               [000212] ------------                 IL_OFFSET void   IL offset: 0x99
N001 (  3,  2) [000095] ------------        t95 =    LCL_VAR   int    V00 loc0         u:7 (last use)
                                                  /--*  t95    int
N003 (  7,  5) [000097] DA----------              *  STORE_LCL_VAR int    V05 loc5         d:2
```
The node is `TYP_INT`, the LLVM type is `i32`, but the value stored in the `_sdsuMap` is an `i8`.  As the `node` is a `TYP_INT`, `varTypeIsSmall(node)` is false and it goes to the "special case for relops" which expects an `i1`

Reply:
Should we `consumeValue` at https://github.com/yowl/runtimelab/blob/b135cf9b031395ffddb1b8479f1f9aef485e43ee/src/coreclr/jit/llvm.cpp#L1619 ?

Reply:
I think the current place is right. We just need to follow some rules on what kinds of type we are to pass as the second parameter (writing them out right now).

Reply:
So, first of all, we have two concepts here:

1) Location types, i. e. those we are to place in the SSA map. These can be small, and must match the local's exact type.
2) Raw tree types. Can be small, exist in the SDSU map, and also can be mismatched with the location types, but only if small.

To maintain the invariant under `1` (only exact types stored in locations), we are to pass the type of the local to `consumeValue` (and ignore the node type).

Now, we also need to decide what location type is proper for a local. This is a little dodgy because the Jit itself is not sure (there are known bugs that exist because of it).

For now, for simplicity purpose, I propose we treat location types as simply `varDsc->TypeGet()`. It should also be used when building the `alloca`s. We could choose to use `TYP_INT` for small-typed locals as well, but it would require more thinking.

This won't solve the assert you observed, where we should have an `i8 -> i32` implicit upcast. This needs another special case (small locals in IR are special) in `consumeValue`: `if node->OperIs(GT_LCL_VAR) { /* use the local's type  */ }`.

Reply:
Thanks, I removed the widening for small ints for the allocas